designs/node-repair.md

```diff
@@ -0,0 +1,92 @@
+#  Karpenter Node Auto Repair 
+
+## Problem Statement
+
+Karpenter users are observing that occasionally nodes can face failure mode that prevent them to be healthy in a Kubernetes clusters. These nodes stay in unhealthy state for a period of time without any actions, and in some cases forever with manual intervention. Karpenter continues to simulate pods to these nodes, but the Kubernetes scheduler won't schedule to them. This will leave pods unscheduled and cause downtime for applications. Customers have the need for karpenter to repair unhealthy nodes. The karpenter team has seen several failure mode such as disk pressure, memory pressure, networking outage etc. that can cause nodes to get into an unhealthy state Here are issues related to the topic:
+
+* Mega Issue: https://github.com/kubernetes-sigs/karpenter/issues/750
+    * Related (Unreachable): https://github.com/aws/karpenter-provider-aws/issues/2570
+    * Related (Remove by taints): https://github.com/aws/karpenter-provider-aws/issues/2544
+    * Related (Known resource are not registered) Fixed by v0.28.0: https://github.com/aws/karpenter-provider-aws/issues/3794
+    * Related (Stuck on NotReady): https://github.com/aws/karpenter-provider-aws/issues/2439
+
+#### Out of scope
+
+The scope of the solution suggested bellow will be an opinionated method that will not give customer configurability. The team does not have enough data to determine the right level of configuration surface that users will utilize. The implementation will not consider budgets, API surface for repairing unhealthy nodes, or option for nodes to be rebooted instead of replaced. The solution will also not give any Karpenter based approach to enable AZ resiliency. **The feature will be gated under an alpha feature flag to allow for additional feedback from customers, and make subsequent changes in the future.**
```

Comment:
```suggestion
The scope of the solution suggested below will be an opinionated method that will not give customer configurability. The team does not have enough data to determine the right level of configuration surface that users will utilize. The implementation will not consider budgets, API surface for repairing unhealthy nodes, or option for nodes to be rebooted instead of replaced. The solution will also not give any Karpenter based approach to enable AZ resiliency. **The feature will be gated under an alpha feature flag to allow for additional feedback from customers, and make subsequent changes in the future.**
```